Subject: [PATCH] On Linux, link against real libunwind
---
Index: runtimes/libunwind/BUILD.bazel
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/runtimes/libunwind/BUILD.bazel b/runtimes/libunwind/BUILD.bazel
--- a/runtimes/libunwind/BUILD.bazel	(revision e306975d579cd2b5c798345fd0fa3baa222c07d8)
+++ b/runtimes/libunwind/BUILD.bazel	(revision e7689c81b7da2cd711e89cb7fdaf49729eea3e07)
@@ -32,11 +32,16 @@

 copy_to_directory(
     name = "libunwind_library_search_directory",
-    srcs = [
-        ":unwind",
-    ] + select({
+    srcs = select({
+        "@platforms//os:linux": [":libunwind.static"],
+        "//conditions:default": [":unwind"],
+    }) + select({
         ":stub_libgcc_s": [":gcc_s"],
         "//conditions:default": [],
     }),
+    include_external_repositories = ["*libunwind"],
+    replace_prefixes = {
+        "libunwind.static_/liblibunwind.static.a": "libunwind.a",
+    },
     visibility = ["//visibility:public"],
 )
