Subject: [PATCH] Disable `llvm_cov` and `llvm_profdata` (not available in `21.1.8-5`)
---
Index: rs/toolchains/declare_rustc_toolchains.bzl
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/rs/toolchains/declare_rustc_toolchains.bzl b/rs/toolchains/declare_rustc_toolchains.bzl
--- a/rs/toolchains/declare_rustc_toolchains.bzl	(revision 0ee91c13ef8bf9583b8c8d007d1467dbbc7db0ac)
+++ b/rs/toolchains/declare_rustc_toolchains.bzl	(revision e4a68363ff63916367dbf9c4c8a291f8b3458a8d)
@@ -58,8 +58,8 @@
             cargo = "{}cargo".format(cargo_repo_label),
             clippy_driver = "{}clippy_driver_bin".format(clippy_repo_label),
             cargo_clippy = "{}cargo_clippy_bin".format(clippy_repo_label),
-            llvm_cov = "@llvm//tools:llvm-cov",
-            llvm_profdata = "@llvm//tools:llvm-profdata",
+            # llvm_cov = "@llvm//tools:llvm-cov",
+            # llvm_profdata = "@llvm//tools:llvm-profdata",
             rust_objcopy = "{}rust-objcopy".format(rustc_repo_label),
             rustc_lib = "{}rustc_lib".format(rustc_repo_label),
             allocator_library = None,
