Subject: [PATCH] Add support for MSVC ABI from `@llvm`
---
Index: rs/platforms/triples.bzl
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/rs/platforms/triples.bzl b/rs/platforms/triples.bzl
--- a/rs/platforms/triples.bzl	(revision 597eefe2a71e192799863d739d0ecef29d866e22)
+++ b/rs/platforms/triples.bzl	(revision 0ee91c13ef8bf9583b8c8d007d1467dbbc7db0ac)
@@ -10,6 +10,12 @@
     "msvc": "@rules_rs//rs/platforms/constraints:windows_msvc",
 }

+_WINDOWS_LLVM_ABI_CONSTRAINTS = {
+    "gnu": "@llvm//constraints/abi:gnu",
+    "gnullvm": "@llvm//constraints/abi:gnullvm",
+    "msvc": "@llvm//constraints/abi:msvc",
+}
+
 def triple_to_constraint_set(target_triple):
     constraints = _triple_to_constraint_set(target_triple)
     t = triple(target_triple)
@@ -23,6 +29,7 @@
             constraints.append("@llvm//constraints/libc:gnu.2.28")
     elif t.system == "windows":
         constraints.append(_WINDOWS_ABI_CONSTRAINTS[t.abi])
+        constraints.append(_WINDOWS_LLVM_ABI_CONSTRAINTS[t.abi])

     return constraints

