Subject: [PATCH] Upgrade `@llvm`
Add support for MSVC ABI from `@llvm`
---
Index: rs/experimental/platforms/triples.bzl
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/rs/experimental/platforms/triples.bzl b/rs/experimental/platforms/triples.bzl
--- a/rs/experimental/platforms/triples.bzl	(revision 5651f1ad10471cd2274034e4f82e046583513c0e)
+++ b/rs/experimental/platforms/triples.bzl	(revision 2e1e6e8418e859aae3d1a5fba5a72c420575dfcd)
@@ -10,6 +10,12 @@
     "msvc": "@rules_rs//rs/experimental/platforms/constraints:windows_msvc",
 }

+_WINDOWS_LLVM_ABI_CONSTRAINTS = {
+    "gnu": "@llvm//constraints/abi:gnu",
+    "gnullvm": "@llvm//constraints/abi:gnullvm",
+    "msvc": "@llvm//constraints/abi:msvc",
+}
+
 def triple_to_constraint_set(target_triple):
     constraints = _triple_to_constraint_set(target_triple)
     t = triple(target_triple)
@@ -23,6 +29,7 @@
             constraints.append("@llvm//constraints/libc:gnu.2.28")
     elif t.system == "windows":
         constraints.append(_WINDOWS_ABI_CONSTRAINTS[t.abi])
+        constraints.append(_WINDOWS_LLVM_ABI_CONSTRAINTS[t.abi])

     return constraints

Index: MODULE.bazel
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/MODULE.bazel b/MODULE.bazel
--- a/MODULE.bazel	(revision 2e1e6e8418e859aae3d1a5fba5a72c420575dfcd)
+++ b/MODULE.bazel	(revision 35f7c253814ebf9538ca56d7c50f2ad760bd9ea8)
@@ -54,7 +54,7 @@
 bazel_dep(name = "protobuf", version = "34.0.bcr.1", repo_name = "com_google_protobuf")
 bazel_dep(name = "rules_cc", version = "0.2.8")
 bazel_dep(name = "rules_proto", version = "7.1.0")
-bazel_dep(name = "llvm", version = "0.7.0")
+bazel_dep(name = "llvm", version = "0.7.1")
 bazel_dep(name = "aspect_tools_telemetry", version = "0.3.2")

 tel = use_extension("@aspect_tools_telemetry//:extension.bzl", "telemetry")
